### DISCUSSION TITLE
Starting and stopping

### DISCUSSION THREAD

**Discussion by 4jag** [inventree/InvenTree#2798] [Q&A] [answered]
What should I do when containers are stuck "restarting"?  I'm sorry if this is really a docker question vs Inventree.

I'm testing in WSL 2 & Docker-Desktop. (It's also a personal workstation, so I need to shut it down restart frequently).

When I first got Inventree running for the first time looks like the first PS below. 
Since Docker-desktop's UI doesn't refresh the list of images and containers, I restarted docker desktop. But this took down the whole docker environment. I'm trying to restart with `docker-compose down` , `docker-compose up -d` but the state of worker and server remain "Restarting..."

How would I tear down the containers and start over, and 
What is the normal way of shutting down and restarting when using Docker-desktop. 
Also does anyone know how to force Docker-desktop into refreshing it's UI data?

```
(env) ~/inventree/docker$ docker ps
CONTAINER ID   IMAGE                        COMMAND                  CREATED         STATUS         PORTS                     NAMES
2e48177a46c8   nginx:stable                 "/docker-entrypoint.…"   2 minutes ago   Up 2 minutes   0.0.0.0:1337->80/tcp      inventree-proxy
cbcd68e2be54   inventree/inventree:stable   "/bin/bash ../docker…"   2 minutes ago   Up 2 minutes                             inventree-worker
ff4a08611aae   inventree/inventree:stable   "/bin/bash ../docker…"   2 minutes ago   Up 2 minutes   8000/tcp                  inventree-server
76dad38da0e6   postgres:13                  "docker-entrypoint.s…"   2 minutes ago   Up 2 minutes   0.0.0.0:62546->5432/tcp   inventree-db
```
Upon restarting it looks like this:
```
(env) ~/inventree/docker$ docker ps
CONTAINER ID   IMAGE                        COMMAND                  CREATED         STATUS                          PORTS                     NAMES
03f2ab13b5df   inventree/inventree:stable   "/bin/bash ../docker…"   5 minutes ago   Restarting (1) 29 seconds ago                             inventree-worker
81fb81c2b007   nginx:stable                 "/docker-entrypoint.…"   5 minutes ago   Up 5 minutes                    0.0.0.0:1337->80/tcp      inventree-proxy
c35900afc160   inventree/inventree:stable   "/bin/bash ../docker…"   5 minutes ago   Restarting (1) 31 seconds ago                             inventree-server
3aaf401d296e   postgres:13                  "docker-entrypoint.s…"   5 minutes ago   Up 5 minutes                    0.0.0.0:62703->5432/tcp   inventree-db
```

NGINX Port question:
Lastly, when I first ran docker-compose run inventree-server invoke update I did so with nginx.conf set with `server {  listen 8022;` since I run a lot of other projects on this workstation and ports 80, 8000 are in use. When I started Inventree I didn't get a response on 8022. I returned it to listen:80 to see if I get something, but by now I have the 'Restarting' issue above going on. ?

Thanks

**Comment by SchrodingersGat** [comment, marked as answer]
Regarding the restarting issue - can you grab the logs from the server and worker containers? There must be something going on there.

Regarding the nginx issue:

- Did you only change the port number in the nginx config file?
- Did you restart the docker-compose setup after doing this?

**Reply by 4jag** [reply]
The only log folder I see is /home/inventree/log, and there's nothing in it. If you mean inside the container -- I can't get in:
 ```
docker exec -it c35900afc160 /bin/bash
Error response from daemon: Container c35900afc160f41d06d47bf37295fbb723748a26d035fd9a5f74a83bb9a5e8a4 is restarting, wait until the container is running.
```
Yes -- I only changed the port in nginx.conf. I'm a little uncertain on which controls the browser-to-nginx and which niginx-to-inventree servers -- is the LISTEN the correct one? (vs LOCATION PROXY_PASS)
And Yes -- Well, I reran 'docker-compose run inventree-server invoke update' after updating nginx.conf. I'm not supposed to backup beyond that, am I? Should I manually remove the containers and images before recreating them (this command), or do i just rerun this cmd?

**Reply by 4jag** [reply]
Hmmm, I'm actually not sure if I did run invoke update after changing nginx.conf. But when I try to run it now -- 'docker-compose run inventree-server invoke update' -- I get:
```
(env) ~/inventree/docker$ docker-compose run inventree-server invoke update
Creating network "docker_default" with the default driver
Creating inventree-db ... done
Creating docker_inventree-server_run ... done
Creating directory /home/inventree/data/static
mkdir: can't create directory '/home/inventree/data/static': Permission denied
ERROR: 1
```

I can create folders in /home/inventree as the current user:
```
(env) ~/inventree/docker$ mkdir /home/inventree/testing
(env) ~/inventree/docker$ ls -l ../..
total 28
drwxrwxr-x 4 inventree inventree 4096 Apr  4 15:32 data
drwxrwxr-x 6 inventree inventree 4096 Mar 15 20:30 env
drwxrwxr-x 9 inventree inventree 4096 Apr  1 16:49 inventree
drwxrwxr-x 2 inventree inventree 4096 Mar 15 20:28 log
drwxrwxr-x 2 inventree inventree 4096 Mar 15 20:28 static
drwxrwxr-x 2 inventree inventree 4096 Apr  4 15:32 test-folder-create
```

**Reply by SchrodingersGat** [reply]
Looks like the external volume mounted for storing media / static data is not being mounted with the correct permissions. 

Can you share the docker-compose file showing where you are mounting the volume from?

**Reply by 4jag** [reply]
I guess I hadn't fixed the "/path/to/data" string. I changed it to "/home/inventree/data", but I got this error:
```
(env) ~/inventree/docker$ docker-compose run inventree-server invoke update
ERROR: Configuration for volume data specifies "device" driver_opt /home/inventree/data, but a volume with the same name uses a different "device" driver_opt (/path/to/data). If you wish to use the new configuration, please remove the existing volume "docker_data" first: $ docker volume rm docker_data
```
Oops. I had to delete all the containers and images, then the docker_data volume. The start again.

Ok, backed up to the very beginning, and modified ONLY docker-compose.yml (put "/home/inventree/data in where it said I have to edit "/path/to.data), and prod-config.env (added my proxy var's). 
I also stopped and deleted all the containers and images, and found I needed to delete the docker_data volume too.
```
(env) ~/inventree/docker$ docker-compose up -d inventree-db
Creating network "docker_default" with the default driver
Pulling inventree-db (postgres:13)...
13: Pulling from library/postgres
c229119241af: Pull complete
...
ba6456490b54: Pull complete
Digest: sha256:43568014f86275fd51992b9c8a6cf528d3acb6da8bcabfac04ed5abecb56a1c1
Status: Downloaded newer image for postgres:13
Creating inventree-db ... done
(env) ~/inventree/docker$ docker-compose run inventree-server pgcli -h inventree-db -p 5432 -u pguser
Creating docker_inventree-server_run ... done
Copying config file to /home/inventree/data/config.yaml
cp: can't create '/home/inventree/data/config.yaml': Permission denied       <------
ERROR: 1

(env) ~/inventree/docker$ ls -l /home/inventree/data/
total 12
drwxrwxr-x  4 inventree inventree 4096 Mar 15 21:55 media
drwx------ 19       999 root      4096 Apr  4 20:27 pgdb
drwxrwxr-x 19 inventree inventree 4096 Mar 15 21:55 static
(env) ~/inventree/docker$ ls -ld /home/inventree/data/                        <------
drwxrwxr-x 5 inventree inventree 4096 Apr  4 20:27 /home/inventree/data/      <------
```
**EDIT**: Turns out deleting this data folder completely fixed the error -- perhaps either the existence of the folder or that it had an old root-owned pgdb in it caused the error?

**Reply by SchrodingersGat** [reply]
The only issue I can see is that the docker container does not have the correct permissions to write to the mounted volume. 

Can you show the portion of the docker-compose file where you configure the volume?

**Reply by 4jag** [reply]
I got it running by deleting the inventree/data folder, getting me past where the script was copying config.yaml into data folder -- Is the script failing on trying to create data if it already exists?
EDIT: There was a root-owned pgdb file (postgres?) in the data folder; prob from an earlier attempt to build a dev instance)

**Reply by 4jag** [reply]
Do you mean this?
```
volumes:
    # NOTE: Change /path/to/data to a directory on your local machine
    # Persistent data, stored external to the container(s)
    data:
        driver: local
        driver_opts:
            type: none
            o: bind
            # This directory specified where InvenTree data are stored "outside" the docker containers
            # Change this path to a local system path where you want InvenTree data stored
            device: /home/inventree/data
```

**Reply by SchrodingersGat** [reply]
So the path specified under "device" must refer to your local (computer) filesystem.. Do you have a local directory called `/home/inventree/data`?

Note; this is *different* to the `/home/inventree/data` internal to the docker container, and does not have to be at that specific path

**Reply by 4jag** [reply]
This in on WSL 2 Ubuntu on a Windows10 PC. So this would mean `/home/inventree/data` exists in the WSL 2 environment, right? (not the Win10 PC). If yes, then that's the ubuntu account running the docker stuff. 
I'm not sure what caused the permission denied errors since the data and static folder were writable and there wasn't a config.yaml file already in it, but for some reason root owned /home/inventree/data/pgdb, and after deleting all the cruft in user inventree's home dir I was able to rebuild the production docker load. (I must have attempted a development build in this folder prior, since the was a "src" folder tree there.

**Reply by SchrodingersGat** [reply]
![image](https://user-images.githubusercontent.com/10080325/161753874-3d3ebd92-1d90-42b6-9e0b-514ba8623b2a.png)

This "device" path *must* point to a writeable directory on your Win10 PC

**Reply by 4jag** [reply]
When I say /home/inventree I'm referring to the WSL user's folder, _not a folder inside the container_. I suspect that inside the inventree-server container you're using the same username "inventree" ...
```
inventree:~$ docker exec -it 4a69dc9aa7e1 /bin/bash
bash-5.1$ pwd
/home/inventree/InvenTree
bash-5.1$ echo $HOME
/home/inventree
bash-5.1$
```
Yup -- I should have made user daffyduck to be less confusing (sorry, I had first followed instructions for building a dev server, so named my WSL user "inventree"). So, now it's just a coincidence that I'm running docker _outside_ in WSL as the same usernames that are _inside_ the server containers.

**Comment by 4jag** [comment]
Ok so knowing the container IDs:
```
$ docker ps
CONTAINER ID   IMAGE                        COMMAND                  CREATED        STATUS          PORTS                     NAMES
1cc634f0cba7   nginx:stable                 "/docker-entrypoint.…"   10 hours ago   Up 34 seconds   0.0.0.0:1337->80/tcp      inventree-proxy
593ad334a3ed   inventree/inventree:stable   "/bin/bash ../docker…"   10 hours ago   Up 39 seconds                             inventree-worker
4a69dc9aa7e1   inventree/inventree:stable   "/bin/bash ../docker…"   10 hours ago   Up 40 seconds   8000/tcp                  inventree-server
30456f6693c9   postgres:13                  "docker-entrypoint.s…"   10 hours ago   Up 36 seconds   0.0.0.0:57065->5432/tcp   inventree-db
```

I can shutdown the system via:
```
$ docker stop 4a69dc9aa7e1 593ad334a3ed 30456f6693c9 1cc634f0cba7
4a69dc9aa7e1
593ad334a3ed
30456f6693c9
1cc634f0cba7
```

Then after quiting Docker-desktop and shutting-down/reboot the PC for a while, restart by Starting Docker-desktop, then:
```
$ docker start 4a69dc9aa7e1 593ad334a3ed 30456f6693c9 1cc634f0cba7
4a69dc9aa7e1
593ad334a3ed
30456f6693c9
1cc634f0cba7
```
